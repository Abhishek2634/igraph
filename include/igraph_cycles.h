
#ifndef IGRAPH_CYCLES_H
#define IGRAPH_CYCLES_H

#include "igraph_constants.h"
#include "igraph_datatype.h"
#include "igraph_decls.h"
#include "igraph_error.h"
#include "igraph_types.h"
#include "igraph_vector_list.h"

__BEGIN_DECLS

IGRAPH_EXPORT igraph_error_t igraph_fundamental_cycles(
    const igraph_t *graph, igraph_vector_int_list_t *result,
    igraph_integer_t start_vid, igraph_integer_t bfs_cutoff,
    const igraph_vector_t *weights);

IGRAPH_EXPORT igraph_error_t igraph_minimum_cycle_basis(
    const igraph_t *graph, igraph_vector_int_list_t *result,
    igraph_integer_t bfs_cutoff, igraph_bool_t complete,
    igraph_bool_t use_cycle_order, const igraph_vector_t *weights);

struct igraph_simple_cycle_search_state_t;
typedef struct igraph_simple_cycle_search_state_t
    igraph_simple_cycle_search_state_t;

IGRAPH_EXPORT igraph_error_t igraph_simple_cycle_search_state_init(
    igraph_simple_cycle_search_state_t *state, const igraph_t *graph);

IGRAPH_EXPORT void igraph_simple_cycle_search_state_destroy(
    igraph_simple_cycle_search_state_t *state);

IGRAPH_EXPORT igraph_error_t igraph_simple_cycles_search_from_one_vertex(
    igraph_simple_cycle_search_state_t *state, igraph_integer_t start,
    igraph_vector_int_list_t *v_result, igraph_vector_int_list_t *e_result,
    igraph_integer_t max_cycle_length);

IGRAPH_EXPORT igraph_error_t igraph_simple_cycles_search_all(
    const igraph_t *graph, igraph_vector_int_list_t *v_result,
    igraph_vector_int_list_t *e_result, igraph_integer_t max_cycle_length);

IGRAPH_EXPORT igraph_error_t igraph_find_cycle(const igraph_t *graph,
                                               igraph_vector_int_t *vertices,
                                               igraph_vector_int_t *edges,
                                               igraph_neimode_t mode);

__END_DECLS

#endif
