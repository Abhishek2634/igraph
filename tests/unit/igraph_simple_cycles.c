/*
   IGraph library.
   Copyright (C) 2021  The igraph development team <igraph@igraph.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <igraph.h>
#include "igraph_cycles.h"
#include "test_utilities.h"
#include <stdlib.h>

void check_cycles(const igraph_t *graph, igraph_integer_t expected) {
    igraph_vector_int_list_t results_v;
    igraph_vector_int_list_t results_e;
    igraph_integer_t i;

    igraph_vector_int_list_init(&results_v, 0);
    igraph_vector_int_list_init(&results_e, 0);

    igraph_simple_cycles_search_all(graph, &results_v, &results_e);

    printf("Finished search, found %" IGRAPH_PRId " cycles.\n\n", igraph_vector_int_list_size(&results_v));

    if (igraph_vcount(graph) < 100) {
        printf("Vertex IDs in cycles:\n");
        print_vector_int_list(&results_v);

        printf("Edge IDs in cycles:\n");
        print_vector_int_list(&results_e);
    }

    for (i = 0; i < expected; i++) {
        igraph_vector_int_t *vertices, *edges;

        vertices = igraph_vector_int_list_get_ptr(&results_v, i);
        edges = igraph_vector_int_list_get_ptr(&results_e, i);
        IGRAPH_ASSERT(igraph_vector_int_size(vertices) == igraph_vector_int_size(edges));
    }

    IGRAPH_ASSERT(igraph_vector_int_list_size(&results_v) == expected);
    IGRAPH_ASSERT(igraph_vector_int_list_size(&results_e) == expected);

    igraph_vector_int_list_destroy(&results_v);
    igraph_vector_int_list_destroy(&results_e);
}

int main(void) {
    igraph_t g, g_ring_undirected, g_star_undirected;

    printf("Testing null graph\n");
    igraph_small(&g, 0, IGRAPH_UNDIRECTED, -1);
    check_cycles(&g, 0);
    igraph_destroy(&g);

    printf("\nTesting empty graph\n");
    igraph_small(&g, 5, IGRAPH_UNDIRECTED, -1);
    check_cycles(&g, 0);
    igraph_destroy(&g);

    printf("\nTesting ring\n");
    igraph_ring(&g, 10, /*directed=*/ 1, /*mutual=*/ 0, /*circular=*/ 1);
    check_cycles(&g, 1);
    igraph_destroy(&g);

    printf("\nTesting large ring\n");
    igraph_ring(&g, 10000, /*directed=*/ 1, /*mutual=*/ 0, /*circular=*/ 1);
    check_cycles(&g, 1);
    igraph_destroy(&g);

    printf("\nTesting star\n");
    igraph_star(&g, 7, IGRAPH_STAR_OUT, 1);
    check_cycles(&g, 0);
    igraph_destroy(&g);

    printf("\nTesting directed wheel\n");
    igraph_wheel(&g, 10, IGRAPH_WHEEL_OUT, 0);
    check_cycles(&g, 1);
    igraph_destroy(&g);

    printf("\nTesting undirected ring\n");
    igraph_ring(&g_ring_undirected, 10, /*directed=*/ 0, /*mutual=*/ 0, /*circular=*/ 1);
    check_cycles(&g_ring_undirected, 1);

    igraph_star(&g_star_undirected, 7, IGRAPH_STAR_UNDIRECTED, 1);
    printf("\nTesting undirected star\n");
    check_cycles(&g_star_undirected, 0);

    igraph_disjoint_union(&g, &g_ring_undirected, &g_star_undirected);
    printf("\nTesting union of undirected wheel and star\n");
    check_cycles(&g, 1);

    printf("\nTesting union of undirected wheel, star and a single edge\n");
    igraph_add_edge(&g, 7, 13); // add a random edge between the two structures to make them connected
    check_cycles(&g, 1);
    igraph_destroy(&g);
    igraph_destroy(&g_star_undirected);
    igraph_destroy(&g_ring_undirected);

    /*
     * This test is temporarily disabled until we sort out the issues with the
     * smaller test cases
     *
    printf("\nTesting undirected wheel\n");
    igraph_wheel(&g, 10, IGRAPH_WHEEL_UNDIRECTED, 0);
    // call cycles finder, expect 65 cycles to be found (
    // 1 cycle of 10 nodes,
    // 10 cycles of 9 nodes,
    // 9 cycles of 8 nodes,
    // 9 cycles of 7 nodes,
    // 9 cycles of 6 nodes,
    // 9 cycles of 5 nodes,
    // 9 cycles of 4 nodes
    // 9 cycles of 3 nodes,
    // )
    check_cycles(&g, 65);
    igraph_destroy(&g);
    */

    igraph_t g_wheel_undirected_2;
    igraph_wheel(&g_wheel_undirected_2, 10, IGRAPH_WHEEL_UNDIRECTED, 0);
    printf("\nCreated undirected wheel\n");
    // call cycles finder, expect 73 cycles to be found (
    // 9 cycle of 10 nodes,
    // 10 cycles of 9 nodes,
    // 9 cycles of 8 nodes,
    // 9 cycles of 7 nodes,
    // 9 cycles of 6 nodes,
    // 9 cycles of 5 nodes,
    // 9 cycles of 4 nodes
    // 9 cycles of 3 nodes,
    // )
    check_cycles(&g_wheel_undirected_2, 73);
    // clean up
    igraph_destroy(&g_wheel_undirected_2);

    ////////////////////////////////
    // Tests as requested in https://github.com/igraph/igraph/pull/2181#issuecomment-1326064152
    /*
     * This graph looks like:
     *
     * 1--2\
     * |  | |
     * 4--3/
     *
     */
    printf("\nTesting directed graph with a cycle of length 4 and a multi-edge\n");
    igraph_small(&g, 5, IGRAPH_DIRECTED, 1, 2, 2, 3, 2, 3, 3, 4, 4, 1, -1);
    check_cycles(&g, 2);
    igraph_destroy(&g);

    // same, but undirected
    printf("\nTesting undirected graph with a cycle of length 4 and a multi-edge\n");
    igraph_small(&g, 5, IGRAPH_UNDIRECTED, 1, 2, 2, 3, 2, 3, 3, 4, 4, 1, -1);
    check_cycles(&g, 2);
    igraph_destroy(&g);

    // check that self-loops are handled
    printf("\nTesting graph with single self-loop\n");
    igraph_small(&g, 1, IGRAPH_DIRECTED, 0, 0, -1);
    check_cycles(&g, 1);
    igraph_destroy(&g);

    ////////////////////////////////

    // clean up test
    VERIFY_FINALLY_STACK();

    return 0;
}
