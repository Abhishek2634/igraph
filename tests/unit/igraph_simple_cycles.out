Testing null graph
Finished search, found 0 cycles.

Vertex IDs in cycles:
{
}
Edge IDs in cycles:
{
}

Testing empty graph
Finished search, found 0 cycles.

Vertex IDs in cycles:
{
}
Edge IDs in cycles:
{
}

Testing ring
Finished search, found 1 cycles.

Vertex IDs in cycles:
{
  0: ( 0 1 2 3 4 5 6 7 8 9 )
}
Edge IDs in cycles:
{
  0: ( 0 1 2 3 4 5 6 7 8 9 )
}

Testing large ring
Finished search, found 1 cycles.


Testing star
Finished search, found 0 cycles.

Vertex IDs in cycles:
{
}
Edge IDs in cycles:
{
}

Testing directed wheel
Finished search, found 1 cycles.

Vertex IDs in cycles:
{
  0: ( 1 2 3 4 5 6 7 8 9 )
}
Edge IDs in cycles:
{
  0: ( 9 10 11 12 13 14 15 16 17 )
}

Testing undirected ring
Finished search, found 1 cycles.

Vertex IDs in cycles:
{
  0: ( 0 1 2 3 4 5 6 7 8 9 )
}
Edge IDs in cycles:
{
  0: ( 0 1 2 3 4 5 6 7 8 9 )
}

Testing undirected star
Finished search, found 0 cycles.

Vertex IDs in cycles:
{
}
Edge IDs in cycles:
{
}

Testing union of undirected wheel and star
Finished search, found 1 cycles.

Vertex IDs in cycles:
{
  0: ( 0 1 2 3 4 5 6 7 8 9 )
}
Edge IDs in cycles:
{
  0: ( 0 1 2 3 4 5 6 7 8 9 )
}

Testing union of undirected wheel, star and a single edge
Finished search, found 1 cycles.

Vertex IDs in cycles:
{
  0: ( 0 1 2 3 4 5 6 7 8 9 )
}
Edge IDs in cycles:
{
  0: ( 0 1 2 3 4 5 6 7 8 9 )
}

Created undirected wheel
Finished search, found 73 cycles.

Vertex IDs in cycles:
{
  0: ( 0 1 2 )
  1: ( 0 1 2 3 )
  2: ( 0 1 2 3 4 )
  3: ( 0 1 2 3 4 5 )
  4: ( 0 1 2 3 4 5 6 )
  5: ( 0 1 2 3 4 5 6 7 )
  6: ( 0 1 2 3 4 5 6 7 8 )
  7: ( 0 1 2 3 4 5 6 7 8 9 )
  8: ( 0 1 9 )
  9: ( 0 1 9 8 )
  10: ( 0 1 9 8 7 )
  11: ( 0 1 9 8 7 6 )
  12: ( 0 1 9 8 7 6 5 )
  13: ( 0 1 9 8 7 6 5 4 )
  14: ( 0 1 9 8 7 6 5 4 3 )
  15: ( 0 1 9 8 7 6 5 4 3 2 )
  16: ( 0 2 1 9 )
  17: ( 0 2 1 9 8 )
  18: ( 0 2 1 9 8 7 )
  19: ( 0 2 1 9 8 7 6 )
  20: ( 0 2 1 9 8 7 6 5 )
  21: ( 0 2 1 9 8 7 6 5 4 )
  22: ( 0 2 1 9 8 7 6 5 4 3 )
  23: ( 0 2 3 )
  24: ( 0 2 3 4 )
  25: ( 0 2 3 4 5 )
  26: ( 0 2 3 4 5 6 )
  27: ( 0 2 3 4 5 6 7 )
  28: ( 0 2 3 4 5 6 7 8 )
  29: ( 0 2 3 4 5 6 7 8 9 )
  30: ( 0 3 2 1 9 )
  31: ( 0 3 2 1 9 8 )
  32: ( 0 3 2 1 9 8 7 )
  33: ( 0 3 2 1 9 8 7 6 )
  34: ( 0 3 2 1 9 8 7 6 5 )
  35: ( 0 3 2 1 9 8 7 6 5 4 )
  36: ( 0 3 4 )
  37: ( 0 3 4 5 )
  38: ( 0 3 4 5 6 )
  39: ( 0 3 4 5 6 7 )
  40: ( 0 3 4 5 6 7 8 )
  41: ( 0 3 4 5 6 7 8 9 )
  42: ( 0 4 3 2 1 9 )
  43: ( 0 4 3 2 1 9 8 )
  44: ( 0 4 3 2 1 9 8 7 )
  45: ( 0 4 3 2 1 9 8 7 6 )
  46: ( 0 4 3 2 1 9 8 7 6 5 )
  47: ( 0 4 5 )
  48: ( 0 4 5 6 )
  49: ( 0 4 5 6 7 )
  50: ( 0 4 5 6 7 8 )
  51: ( 0 4 5 6 7 8 9 )
  52: ( 0 5 4 3 2 1 9 )
  53: ( 0 5 4 3 2 1 9 8 )
  54: ( 0 5 4 3 2 1 9 8 7 )
  55: ( 0 5 4 3 2 1 9 8 7 6 )
  56: ( 0 5 6 )
  57: ( 0 5 6 7 )
  58: ( 0 5 6 7 8 )
  59: ( 0 5 6 7 8 9 )
  60: ( 0 6 5 4 3 2 1 9 )
  61: ( 0 6 5 4 3 2 1 9 8 )
  62: ( 0 6 5 4 3 2 1 9 8 7 )
  63: ( 0 6 7 )
  64: ( 0 6 7 8 )
  65: ( 0 6 7 8 9 )
  66: ( 0 7 6 5 4 3 2 1 9 )
  67: ( 0 7 6 5 4 3 2 1 9 8 )
  68: ( 0 7 8 )
  69: ( 0 7 8 9 )
  70: ( 0 8 7 6 5 4 3 2 1 9 )
  71: ( 0 8 9 )
  72: ( 1 2 3 4 5 6 7 8 9 )
}
Edge IDs in cycles:
{
  0: ( 0 9 1 )
  1: ( 0 9 10 2 )
  2: ( 0 9 10 11 3 )
  3: ( 0 9 10 11 12 4 )
  4: ( 0 9 10 11 12 13 5 )
  5: ( 0 9 10 11 12 13 14 6 )
  6: ( 0 9 10 11 12 13 14 15 7 )
  7: ( 0 9 10 11 12 13 14 15 16 8 )
  8: ( 0 17 8 )
  9: ( 0 17 16 7 )
  10: ( 0 17 16 15 6 )
  11: ( 0 17 16 15 14 5 )
  12: ( 0 17 16 15 14 13 4 )
  13: ( 0 17 16 15 14 13 12 3 )
  14: ( 0 17 16 15 14 13 12 11 2 )
  15: ( 0 17 16 15 14 13 12 11 10 1 )
  16: ( 1 9 17 8 )
  17: ( 1 9 17 16 7 )
  18: ( 1 9 17 16 15 6 )
  19: ( 1 9 17 16 15 14 5 )
  20: ( 1 9 17 16 15 14 13 4 )
  21: ( 1 9 17 16 15 14 13 12 3 )
  22: ( 1 9 17 16 15 14 13 12 11 2 )
  23: ( 1 10 2 )
  24: ( 1 10 11 3 )
  25: ( 1 10 11 12 4 )
  26: ( 1 10 11 12 13 5 )
  27: ( 1 10 11 12 13 14 6 )
  28: ( 1 10 11 12 13 14 15 7 )
  29: ( 1 10 11 12 13 14 15 16 8 )
  30: ( 2 10 9 17 8 )
  31: ( 2 10 9 17 16 7 )
  32: ( 2 10 9 17 16 15 6 )
  33: ( 2 10 9 17 16 15 14 5 )
  34: ( 2 10 9 17 16 15 14 13 4 )
  35: ( 2 10 9 17 16 15 14 13 12 3 )
  36: ( 2 11 3 )
  37: ( 2 11 12 4 )
  38: ( 2 11 12 13 5 )
  39: ( 2 11 12 13 14 6 )
  40: ( 2 11 12 13 14 15 7 )
  41: ( 2 11 12 13 14 15 16 8 )
  42: ( 3 11 10 9 17 8 )
  43: ( 3 11 10 9 17 16 7 )
  44: ( 3 11 10 9 17 16 15 6 )
  45: ( 3 11 10 9 17 16 15 14 5 )
  46: ( 3 11 10 9 17 16 15 14 13 4 )
  47: ( 3 12 4 )
  48: ( 3 12 13 5 )
  49: ( 3 12 13 14 6 )
  50: ( 3 12 13 14 15 7 )
  51: ( 3 12 13 14 15 16 8 )
  52: ( 4 12 11 10 9 17 8 )
  53: ( 4 12 11 10 9 17 16 7 )
  54: ( 4 12 11 10 9 17 16 15 6 )
  55: ( 4 12 11 10 9 17 16 15 14 5 )
  56: ( 4 13 5 )
  57: ( 4 13 14 6 )
  58: ( 4 13 14 15 7 )
  59: ( 4 13 14 15 16 8 )
  60: ( 5 13 12 11 10 9 17 8 )
  61: ( 5 13 12 11 10 9 17 16 7 )
  62: ( 5 13 12 11 10 9 17 16 15 6 )
  63: ( 5 14 6 )
  64: ( 5 14 15 7 )
  65: ( 5 14 15 16 8 )
  66: ( 6 14 13 12 11 10 9 17 8 )
  67: ( 6 14 13 12 11 10 9 17 16 7 )
  68: ( 6 15 7 )
  69: ( 6 15 16 8 )
  70: ( 7 15 14 13 12 11 10 9 17 8 )
  71: ( 7 16 8 )
  72: ( 9 10 11 12 13 14 15 16 17 )
}

Testing directed graph with a cycle of length 4 and a multi-edge
Finished search, found 2 cycles.

Vertex IDs in cycles:
{
  0: ( 1 2 3 4 )
  1: ( 1 2 3 4 )
}
Edge IDs in cycles:
{
  0: ( 0 2 3 4 )
  1: ( 0 1 3 4 )
}

Testing undirected graph with a cycle of length 4 and a multi-edge
Finished search, found 2 cycles.

Vertex IDs in cycles:
{
  0: ( 1 2 3 4 )
  1: ( 1 2 3 4 )
}
Edge IDs in cycles:
{
  0: ( 0 2 3 4 )
  1: ( 0 1 3 4 )
}

Testing graph with single self-loop
Finished search, found 1 cycles.

Vertex IDs in cycles:
{
  0: ( 0 )
}
Edge IDs in cycles:
{
  0: ( 0 )
}
